src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs

```diff
@@ -0,0 +1,463 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    public partial class UseConcreteTypeAnalyzer
+    {
+        private sealed class Collector : IDisposable
+        {
+            private static readonly ObjectPool<Collector> _pool = new(() => new Collector());
+
+            public readonly ConcurrentDictionary<IFieldSymbol, bool> VirtualDispatchFields = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, bool> VirtualDispatchLocals = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, bool> VirtualDispatchParameters = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, bool> MethodsAssignedToDelegate = new(SymbolEqualityComparer.Default);
+
+            public readonly ConcurrentDictionary<IFieldSymbol, ConcurrentBag<ITypeSymbol>> FieldAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<ILocalSymbol, ConcurrentBag<ITypeSymbol>> LocalAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IParameterSymbol, ConcurrentBag<ITypeSymbol>> ParameterAssignments = new(SymbolEqualityComparer.Default);
+            public readonly ConcurrentDictionary<IMethodSymbol, ConcurrentBag<ITypeSymbol>> MethodReturns = new(SymbolEqualityComparer.Default);
+
+            public INamedTypeSymbol? Void { get; private set; }
+
+            private Collector()
+            {
+            }
+
+            public void Dispose()
+            {
+                VirtualDispatchFields.Clear();
+                VirtualDispatchLocals.Clear();
+                VirtualDispatchParameters.Clear();
+                MethodsAssignedToDelegate.Clear();
+                FieldAssignments.Clear();
+                LocalAssignments.Clear();
+                ParameterAssignments.Clear();
+                MethodReturns.Clear();
+                Void = null;
+            }
+
+            public static Collector GetInstance(Compilation compilation)
+            {
+                var c = _pool.Allocate();
```

Comment:
Where is the collector instance returned back to the pool?

Reply:
You probably need to define the below and invoke that from SymbolEndAction:

```csharp
public static void Free(Collector collector)
{
    collector.Dispose();
    pool.Free(collector);
}
```

Reply:
Whoops. Done.